### PR TITLE
Add elastic buffer stability checker

### PR DESCRIPTION
**Pull request by lmbollen**
An initial implementation for a component that observes the buffer occupancy of an elastic buffer and determines if it is stable. (see #126)

I added the possibility to create multiple strategies to determine whether the buffer occupancies have converged to an equilibrium. This enables us to try multiple different approaches.

TODO:
- [x] ~~Add more strategies?~~
- [x] Add tests that verify that the available strategies work as intended.

### CODE REVIEW

File: bittide/src/Bittide/ClockControl/StabilityChecker.hs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  -- from the current @target@ for it to be considered "stable".
+  SNat margin ->
+  -- | Minimum number of clock cycles the incoming buffer occupancy must remain within the
+  -- @margin@ for it to be considered "stable".
+  SNat cyclesStable ->
+  -- | Incoming buffer occupancy.
+  Signal dom (Unsigned n) ->
+  -- | Stability indicator.
+  Signal dom Bool
+checkTargetDeviation SNat SNat = mealy go Nothing
```
Comment (martijnbastiaan) [inline bittide/src/Bittide/ClockControl/StabilityChecker.hs:52]: If you start with `(0,0)` you can get rid of the `Maybe`/`fromJust`. The state machine will immediately jump to `(0, target)`.

File: bittide/src/Bittide/ClockControl/StabilityChecker.hs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ where
+  go oldState input = (newState, output)
+   where
+    (cnt, target) = fromJust oldState
+    withinMargin = abs (toSigned target - toSigned input) < (natToNum @margin)
```
Comment (martijnbastiaan) [inline bittide/src/Bittide/ClockControl/StabilityChecker.hs:57]:
```suggestion
    withinMargin = abs (toSigned target `sub` toSigned input) < (natToNum @margin)
```

where `sub` comes from [ExtendingNum](https://hackage.haskell.org/package/clash-prelude-1.6.4/docs/Clash-Class-Num.html#t:ExtendingNum). Currently this is not guaranteed to not underflow.

File: bittide/src/Bittide/ClockControl/StabilityChecker.hs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    newState :: Maybe (Index cyclesStable, Unsigned n)
+    newState
+      | isNothing oldState = Just (0, input)
+      | withinMargin       = Just (satSucc SatBound cnt, target)
+      | otherwise          = Just (0, target)
+
+    output = isJust oldState && withinMargin && cnt == maxBound
```
Comment (martijnbastiaan) [inline bittide/src/Bittide/ClockControl/StabilityChecker.hs:65]:
```suggestion
    stable = isJust oldState && withinMargin && cnt == maxBound
```

**Review by martijnbastiaan — CHANGES_REQUESTED** [review]
Some tiny changes requested. I think some static unit tests should cover this. E.g., give it a list of numbers wildly varying at the start, stable at the end, and check whether the algorithm emits `stable ~ true` at the right time.

I don't think we're going to need anything else, so in that sense I feel already accounting for different algos is overengineering, but I won't shoot down the PR over it :-).

File: bittide/src/Bittide/ClockControl/StabilityChecker.hs (hunk 1 above)
Comment (lmbollen) [reply bittide/src/Bittide/ClockControl/StabilityChecker.hs:52]: Perhaps we should just initialize the target at half the range of `Unsigned n`.

Setting it to `(0,0)` would always initialize the target at `0`.

File: bittide/src/Bittide/ClockControl/StabilityChecker.hs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  StabilityStrategy ->
+  -- | Incoming buffer occupancy.
+  Signal dom (Unsigned n) ->
+  -- | Stability indicator.
+  Signal dom Bool
+stabilityChecker = \case
+  TrackedTargetDeviation maxDev cyclesStable ->
+    checkTrackedTargetDeviation maxDev cyclesStable
+
+-- | Checks if the @Signal@ of buffer occupancies from an elastic buffer is stable.
```
Comment (martijnbastiaan) [inline bittide/src/Bittide/ClockControl/StabilityChecker.hs:34]:
```suggestion
-- | Checks whether the @Signal@ of buffer occupancies from an elastic buffer is stable.
```

Also check out: https://site.uit.no/english/grammar/whether-or-if/

File: bittide/src/Bittide/ClockControl/StabilityChecker.hs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+    newState :: (Index (cyclesStable + 1), Unsigned n)
+    newState
+      | withinMargin = (satSucc SatBound cnt, target)
+      | otherwise    = (0, input)
+
+    isStable = withinMargin && cnt == maxBound
+
+  toSigned :: KnownNat n => Unsigned n -> Signed (n+1)
+  toSigned = bitCoerce . zeroExtend
```
Comment (martijnbastiaan) [inline bittide/src/Bittide/ClockControl/StabilityChecker.hs:66]: `unsignedToSigned` already exists in `Bittide.ClockControl.Callisto.Util`. I think we should move that to `Clash.Sized.Extra` and use that here.

File: bittide/src/Bittide/ClockControl/StabilityChecker.hs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  -- from the current @target@ for it to be considered "stable".
+  SNat margin ->
+  -- | Minimum number of clock cycles the incoming buffer occupancy must remain within the
+  -- @margin@ for it to be considered "stable".
+  SNat cyclesStable ->
+  -- | Incoming buffer occupancy.
+  Signal dom (Unsigned n) ->
+  -- | Stability indicator.
+  Signal dom Bool
+checkTrackedTargetDeviation SNat SNat = mealy go (0, shiftR maxBound 1)
```
Comment (martijnbastiaan) [inline bittide/src/Bittide/ClockControl/StabilityChecker.hs:52]:
```suggestion
checkTrackedTargetDeviation SNat SNat = mealy go (0, targetDataCount)
```

Comment (martijnbastiaan) [reply bittide/src/Bittide/ClockControl/StabilityChecker.hs:52]: See `Bittide.ClockControl.targetDataCount`.

File: bittide/tests/Tests/StabilityChecker.hs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+import Clash.Prelude hiding ((^))
+import Test.Tasty
+import qualified Hedgehog.Gen as Gen
+import Hedgehog
+import Bittide.ClockControl.StabilityChecker
+import Data.Proxy
+import Tests.Shared
+import qualified Hedgehog.Range as Range
+import Clash.Hedgehog.Sized.Unsigned (genUnsigned)
+import Test.Tasty.Hedgehog
```
Comment (martijnbastiaan) [inline bittide/tests/Tests/StabilityChecker.hs:19]: sort / move / etc.

File: bittide/tests/Tests/StabilityChecker.hs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  case (someNatVal m, someNatVal (cs - 1)) of
+    (Just (SomeNat margin), Just (SomeNat cyclesStable)) ->
+      TrackedTargetDeviation (snatProxy margin) (succSNat $ snatProxy cyclesStable)
+    _ -> error $
+      "mkTrackedTargetDeviation: ( 0 <= margin) and (1 <= cyclesStable) not met: margin = "
+      <> show m <> ", cyclesStable = " <> show cs <> "."
+
+trackedTargetDeviationTest :: Property
+trackedTargetDeviationTest = property $ do
+  dataCountBits <- forAll $ Gen.integral $ Range.linear 2 63
```
Comment (martijnbastiaan) [inline bittide/tests/Tests/StabilityChecker.hs:37]:
```suggestion
  dataCountBits <- forAll $ Gen.integral $ Range.linear 2 128
```

Cross the machine limit. (Not that we're actually going to use >64 bits to represent data count but you know..)

File: bittide/tests/Tests/StabilityChecker.hs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        topEntity = wcre $ stabilityChecker @System @dataCountBits strategy
+        simOut = simulateN simLength topEntity dataCounts
+
+      simOut === golden margin cyclesStable dataCounts
+    _ -> error "trackedTargetDeviationTest: Initial configuration is invalid."
+ where
+  golden :: forall n . KnownNat n => Integer -> Integer -> [Unsigned n] -> [Bool]
+  golden margin cyclesStable dataCounts =
+    f
+    (0, fromIntegral $ shiftR (maxBound :: Unsigned n) 1)
```
Comment (martijnbastiaan) [inline bittide/tests/Tests/StabilityChecker.hs:55]:
```suggestion
    (0, fromIntegral $ targetDataCount :: Unsigned n)
```

**Review by martijnbastiaan — CHANGES_REQUESTED** [review]

**Comment by martijnbastiaan** [comment]
See PR I referenced just now. Could you add

```
Co-authored-by: Vanessa McHale <vanessa@qbaylogic.com>
```

to the commit body?

File: bittide/src/Bittide/ClockControl/StabilityChecker.hs (hunk 5 above)
Comment (lmbollen) [reply bittide/src/Bittide/ClockControl/StabilityChecker.hs:66]: Say no more

**Review by martijnbastiaan — APPROVED** [review]

**Pull request merged**